### PR TITLE
Add the article about Scalar Quantization

### PR DESCRIPTION
**Pull request by kacperlukawski**

### CODE REVIEW

**Comment by generall** [comment]
> Because of SIMD, we use a shorter i8 range: [0; 127]. In SIMD calculations we have represent quantized data as u8 and i8 both without conversions

SIMD details are out of scope for theory part

**Comment by IvanPleshkov** [comment]
> > Because of SIMD, we use a shorter i8 range: [0; 127]. In SIMD calculations we have represent quantized data as u8 and i8 both without conversions
> 
> SIMD details are out of scope for theory part

But user can skip the detailed description of quantization and start thinking that we use whole [-128;128] range

**Comment by kacperlukawski** [comment]
> > > Because of SIMD, we use a shorter i8 range: [0; 127]. In SIMD calculations we have represent quantized data as u8 and i8 both without conversions
> > 
> > 
> > SIMD details are out of scope for theory part
> 
> But user can skip the detailed description of quantization and start thinking that we use whole [-128;128] range

How about making a separate section about "Qdrant implementation" and putting that info there?

**Comment by IvanPleshkov** [comment]
`memory_size = number_of_vectors * vector_dimension * 4 bytes * 1.5`
I propose to remove word `bytes`. Or move `1.5` to beginning of formula like this:
`memory_size (in bytes) = 1.5 * number_of_vectors * vector_dimension * 4 bytes`

File: qdrant-landing/content/articles/scalar-quantization.md [hunk 1]
```diff
@@ -0,0 +1,269 @@
...
+might be represented. Within a single byte they may represent up to 256 different values,
+either from `[-128, 127]` or `[0, 255]`.
+
+![Value ranges represented by int8](/articles_data/scalar-quantization/int8-value-range.png)
+
+Since we put some boundaries on the numbers that might be represented by the `f32`, and
+`i8` has some natural boundaries, the process of converting the values between those
+two ranges is quite natural:
+
+$$ f32 = \alpha * i8 + offset $$
```
Comment (generall) [inline qdrant-landing/content/articles/scalar-quantization.md:78]:
```suggestion
$$ f32 = \alpha \times i8 + offset $$
```

File: qdrant-landing/content/articles/scalar-quantization.md [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+The parameters $ \alpha $ and $ offset $ has to be calculated for a given set of vectors, 
+but that comes easily by putting the minimum and maximum of the represented range for 
+both `f32` and `i8`. 
+
+![Float32 to int8 conversion](/articles_data/scalar-quantization/float32-to-int8-conversion.png)
+
+For the unsigned `int8` it will go as following:
+
+$$ \begin{equation}
+\begin{cases} -2 = \alpha * 0 + offset \\\\ 5 = \alpha * 255 + offset \end{cases}
```
Comment (generall) [inline qdrant-landing/content/articles/scalar-quantization.md:91]:
```suggestion
\begin{cases} -2 = \alpha * 0 + offset \\\\ 5 = \alpha \times 255 + offset \end{cases} 
```

File: qdrant-landing/content/articles/scalar-quantization.md [hunk 3]
```diff
@@ -0,0 +1,269 @@
...
+For the unsigned `int8` it will go as following:
+
+$$ \begin{equation}
+\begin{cases} -2 = \alpha * 0 + offset \\\\ 5 = \alpha * 255 + offset \end{cases} 
+\end{equation} $$
+
+In case of signed `int8`, we'll just change the represented range boundaries:
+
+$$ \begin{equation}
+\begin{cases} -2 = \alpha * (-128) + offset \\\\ 5 = \alpha * 127 + offset \end{cases}
```
Comment (generall) [inline qdrant-landing/content/articles/scalar-quantization.md:97]:
```suggestion
\begin{cases} -2 = \alpha * (-128) + offset \\\\ 5 = \alpha \times 127 + offset \end{cases} 
```

File: qdrant-landing/content/articles/scalar-quantization.md [hunk 4]
```diff
@@ -0,0 +1,269 @@
...
+We do not store the vectors in the collections represented by `int8` instead of `float32` 
+just for the sake of compressing the memory. But the coordinates are being used while we 
+calculate the distance between the vectors. Both dot product and cosine distance requires 
+multiplying the corresponding coordinates of two vectors, so that's the operation we 
+perform quite often on `float32`. Here is how it would look like if we perform the 
+conversion to `int8`:
+
+$$ f32 * f32' = $$
+$$ = (\alpha * i8 + offset) * (\alpha * i8' + offset) = $$
+$$ = \alpha^{2} * i8 * i8' + offset * \alpha * i8' + offset * \alpha * i8 + offset^{2} $$
```
Comment (generall) [inline qdrant-landing/content/articles/scalar-quantization.md:115]:
```suggestion
$$ f32 \times f32' = $$
$$ = (\alpha \times i8 + offset) \times (\alpha \times i8' + offset) = $$
$$ = \alpha^{2} * i8 * i8' + offset \times \alpha \times i8' + offset \times \alpha \times i8 + offset^{2} $$
```

File: qdrant-landing/content/articles/scalar-quantization.md [hunk 5]
```diff
@@ -0,0 +1,269 @@
...
+
+The first term, $ \alpha^{2} * i8 * i8' $ has to be calculated when we measure the
```
Comment (generall) [inline qdrant-landing/content/articles/scalar-quantization.md:117]:
```suggestion
The first term, $ \alpha^{2} \times i8 \times i8' $ has to be calculated when we measure the
```

File: qdrant-landing/content/articles/scalar-quantization.md [hunk 6]
```diff
@@ -0,0 +1,269 @@
...
+distance as it depends on both vectors. However, both the second and the third term 
+($ offset * \alpha * i8' $ and $ offset * \alpha * i8 $ respectively), depend only on a
```
Comment (generall) [inline qdrant-landing/content/articles/scalar-quantization.md:119]:
```suggestion
($ offset \times \alpha \times i8' $ and $ offset \times \alpha \times i8 $ respectively), depend only on a 
```

File: qdrant-landing/content/articles/scalar-quantization.md [hunk 7]
```diff
@@ -0,0 +1,269 @@
...
+draft: false
+keywords:
+  - vector search
+  - scalar quantization
+  - TBD
+---
+
+High-dimensional vector embeddings can be memory-intensive, especially when working with 
+large datasets consisting of millions of vectors. Memory footprint really starts being 
+a concern when we scale things up. A small choice of the data type used to store a single
```
Comment (generall) [inline qdrant-landing/content/articles/scalar-quantization.md:21]: maybe `small` -> `simple`?

As I understood from https://context.reverso.net/translation/english-polish/A+small+choice

It rather means a small variety of options

File: qdrant-landing/content/articles/scalar-quantization.md [hunk 8]
```diff
@@ -0,0 +1,269 @@
...
+2 kB. That's only the memory used to store the vector. There is also an overhead of the
+HNSW graph, so as a rule of thumb we estimate the memory size with the following formula:
+
+```
+memory_size = number_of_vectors * vector_dimension * 4 bytes * 1.5
+```
+
+While Qdrant offers various options to store some parts of the data on disk, starting
+from the version 1.1.0, you can also optimize your memory by changing the data type used
+by the embeddings. We've implemented the mechanism of **Scalar Quantization**!
```
Comment (generall) [inline qdrant-landing/content/articles/scalar-quantization.md:38]: We are not exactly changing the data type of vectors. We still store original. Maybe we could say that we are compressing vectors?
Also, we improve speed as well, not only memory

**Comment by IvanPleshkov** [comment]
I propose to remove `Upload time` and keep only `Upload + Indexing time`. Because upload time should not change, it's unstable measurement results

File: qdrant-landing/content/articles/scalar-quantization.md [hunk 9]
```diff
@@ -0,0 +1,291 @@
...
+We do not store the vectors in the collections represented by `int8` instead of `float32` 
+just for the sake of compressing the memory. But the coordinates are being used while we 
+calculate the distance between the vectors. Both dot product and cosine distance requires 
+multiplying the corresponding coordinates of two vectors, so that's the operation we 
+perform quite often on `float32`. Here is how it would look like if we perform the 
+conversion to `int8`:
+
+$$ f32 \times f32' = $$
+$$ = (\alpha \times i8 + offset) \times (\alpha \times i8' + offset) = $$
+$$ = \alpha^{2} \times i8 \times i8' + offset \times \alpha \times i8' + offset \times \alpha \times i8 + offset^{2} $$
```
Comment (generall) [inline qdrant-landing/content/articles/scalar-quantization.md:116]:
```suggestion
$$ = \alpha^{2} \times i8 \times i8' + \underbrace{offset \times \alpha \times i8' + offset \times \alpha \times i8 + offset^{2}}_\text{pre-compute} $$```

It should look like this:
![image](https://user-images.githubusercontent.com/1935623/226982084-bc9e38ec-260f-4f85-a064-0659fcc7dbb9.png)

(tried to highlight individual terms as well, but MathJax could not handle this for some reason =( )

**Review by generall — APPROVED** [review]

File: qdrant-landing/content/articles/scalar-quantization.md [hunk 10]
```diff
@@ -0,0 +1,293 @@
...
+conversion to `int8` will surely lose some precision, so we rather prefer keeping the 
+representation accuracy within the range of 99% of the most probable values and ignoring
+the precision of the outliers. There might be a different choice of the range width, 
+actually, any value from a range `[0, 1]`, where `0` means empty range, and `1` would 
+keep all the values. That's a hyperparameter of the procedure called `quantile`. A value 
+of `0.95` or `0.99` is typically a reasonable choice, but in general `quantile ∈ [0, 1]`.
+
+#### Conversion to integers
+
+Let's talk about conversion to `int8`. Integers also have a finite set of values that
```
Comment (agourlay) [inline qdrant-landing/content/articles/scalar-quantization.md:69]:
```suggestion
Let's talk about the conversion to `int8`. Integers also have a finite set of values that
```

File: qdrant-landing/content/articles/scalar-quantization.md [hunk 11]
```diff
@@ -0,0 +1,293 @@
...
+HNSW graph, so as a rule of thumb we estimate the memory size with the following formula:
+
+```
+memory_size = 1.5 * number_of_vectors * vector_dimension * 4 bytes
+```
+
+While Qdrant offers various options to store some parts of the data on disk, starting 
+from version 1.1.0, you can also optimize your memory by compressing the embeddings. 
+We've implemented the mechanism of **Scalar Quantization**! It turns out to have not 
+only a positive impact on memory but also performance.
```
Comment (agourlay) [inline qdrant-landing/content/articles/scalar-quantization.md:39]:
```suggestion
only a positive impact on memory but also on the performance. 
```

File: qdrant-landing/content/articles/scalar-quantization.md [hunk 12]
```diff
@@ -0,0 +1,293 @@
...
+($ offset \times \alpha \times i8' $ and $ offset \times \alpha \times i8 $ respectively), 
+depend only on a single vector and those might be precomputed and kept for each vector. 
+The last term, $ offset^{2} $ does not depend on any of the values, so it might be even 
+computed once and reused.
+
+If we had to calculate all the terms to measure the distance, the performance could have 
+been even worse than without the conversion. But thanks for the fact we can precompute
+the majority of the terms, things are getting simpler. And in turns out the scalar 
+quantization has a positive impact not only on the memory usage, but also on the 
+performance. But we did some benchmarks to back this statement!
```
Comment (agourlay) [inline qdrant-landing/content/articles/scalar-quantization.md:129]:
```suggestion
performance. As usual, we performed some benchmarks to support this statement!
```

**Review by agourlay — APPROVED** [review]
Great job on this article, I learned a lot reading it!

I have only minor syntax nitpicks to report.

**Comment by morsik** [comment]
https://github.com/qdrant/landing_page/issues/1048

**Pull request merged**